tripal/tripal
# Chado storage cleanup

# Tripal 4 Core Dev Task

Issue #1593 


### Tripal Version: 4

## Description



While @spficklin and @Ferrisx4 were working on publishing chado content it became aparant that while property types were being added via addTypes, they were not being used. Instead only property types passed into the values array were being used. This triggered a deep search to ensure that all parts of ChadoStorage were being used as expected.

This PR fixes the following:
- Ensures that property types with the same name from different fields are not overwritten. This resulted in the new method signature for add/getTypes()
- Adds an `operation` key to the values array to support findValues() which will be implemented in another PR
- Removes `definition` and `type` keys from the values array as they are now added to chadostorage via separate methods. This reduces repetition in the values array.
- Updates TripalEntity to use addTypes() and addFieldDefinitions() rather than passing this information into the values array.
- Updates ChadoStorage to pull all type and field definition information using getPropertyType() and getFieldDefinition(). It no longer uses the `type` and `definition` keys in the values array
- Updates function docblocks to match this new structure of the values array
- Removes the unused Drupal\tripal\Entity\TripalEntityStorageSchema which was a relic from times long past.
- Updates existing automated tests to confirm these changes.
- Adds in a base TripalStorage class for ChadoStorage to inherit from to contain the add/getFieldDefinition methods as these should be database independent.

## Testing?




### Automated Testing
This code is tested by both Functional + Kernel tests found in:
- tripal_chado/tests/src/Functional/Plugin/ChadoStorageTest.php
- tripal_chado/tests/src/Kernel/Plugin/ChadoStorage/ChadoLinkerPropertyDefaultTest.php

These tests were pre-existing but have been updated to not include the `definition` and `type` keys in the values array and to use addFieldDefinition(). They already used addTypes() but were updated to use the new method signature.

The Kernel tests in particular are quite thorough testing that addTypes(), getTypes(), addFieldDefinition(), insertValues(), loadValues(), updateValues() all perform as expected. They use the property field on a gene content type as an example right now but more such field tests will be added in the future.

NOTE: We do not yet have automated testing for removeTypes(), validateValues() and the TripalEntity class.

### Manual Testing

1. Build a new docker image based on this branch and start a fresh container.
```
git clone https://github.com/tripal/tripal.git tripal-1593-chadoCleanup
cd tripal-1593-chadoCleanup
git checkout tv4g2-1593-chadoStorageCleanup
docker build --tag=tripaldocker:tripal1593 --build-arg drupalversion=10.0.x-dev --file tripaldocker/Dockerfile-php8.2-pgsql13 ./
docker run --publish=80:80 -tid --name=tripal1593 --volume=/Users/las166/Docker/tripal-1593-chadoStorageCleanup:/var/www/drupal9/web/modules/contrib/tripal tripaldocker:tripal1593
docker exec tripal1593 service postgresql restart
```
2. Login to your new site at localhost
3. Test Various content creation, updating and loading (e.g. organism, gene, Germplasm Variety)
    - go to Tripal > Content in the Admin toolbar
    - Click the "Add Tripal Content" button
    - Select the content type you want to test
    - Fill out the form and click Save
    - Check that the next page loads the information you entered in the form
    - Check the database matches the information you entered in the form. For example, `docker exec -it tripal1593 drush sql:query "SELECT * FROM chado.organism"`
    - Go back to the page and click the "Edit" tab.
    - Change information and click save.
    - Check that the next page loads the updated information from the form.
    - Check the database matches the updated information.
